pkg/bindinfo/binding_cache.go

```diff
@@ -52,26 +56,47 @@ type fuzzyBindingCache struct {
 	fuzzy2SQLDigests map[string][]string // fuzzyDigest --> sqlDigests
 
 	sql2FuzzyDigest map[string]string // sqlDigest --> fuzzyDigest
+
+	// loadBindingFromStorageFunc is used to load binding from storage if cache miss.
+	loadBindingFromStorageFunc func(sqlDigest string) (Bindings, error)
 }
 
-func newFuzzyBindingCache() FuzzyBindingCache {
+func newFuzzyBindingCache(loadBindingFromStorageFunc func(string) (Bindings, error)) FuzzyBindingCache {
 	return &fuzzyBindingCache{
-		BindingCache:     newBindCache(),
-		fuzzy2SQLDigests: make(map[string][]string),
-		sql2FuzzyDigest:  make(map[string]string),
+		BindingCache:               newBindCache(),
+		fuzzy2SQLDigests:           make(map[string][]string),
+		sql2FuzzyDigest:            make(map[string]string),
+		loadBindingFromStorageFunc: loadBindingFromStorageFunc,
 	}
 }
 
 func (fbc *fuzzyBindingCache) FuzzyMatchingBinding(sctx sessionctx.Context, fuzzyDigest string, tableNames []*ast.TableName) (matchedBinding Binding, isMatched bool) {
 	fbc.mu.RLock()
-	defer fbc.mu.RUnlock()
 	bindingCache := fbc.BindingCache
 	if bindingCache.Size() == 0 {
+		fbc.mu.RUnlock()
 		return
 	}
 
 	leastWildcards := len(tableNames) + 1
 	enableFuzzyBinding := sctx.GetSessionVars().EnableFuzzyBinding
+	matchedBinding, leastWildcards, isMatched, missingSQLDigest := fbc.getFromMemory(sctx, bindingCache, fuzzyDigest, tableNames, enableFuzzyBinding, leastWildcards)
+	if len(missingSQLDigest) == 0 {
+		fbc.mu.RUnlock()
+		return
+	}
+	fbc.mu.RUnlock()
+	if fbc.loadBindingFromStorageFunc == nil {
+		return
+	}
+	fbc.loadFromStore(missingSQLDigest) // loadFromStore's SetBinding has a Mutex inside, so it's safe to call it without lock
+	fbc.mu.RLock()
+	defer fbc.mu.RUnlock()
```

Comment:
It seems like too many `Lock` and `Unlock`. Can we move these 2 operations into `getFromMeory`? Then this function can be very concise:
```
func (fbc) FuzzyMatchingBinding(...) {
    bindings, ..., missingBindings := fbc.getFromMemory(...)
    if missingBindings == nil {return}
    fbc.loadFromStore(missingBindings)
    return fbc.getFromMemory(...)
}
```
No any lock and unlock operation.

Reply:
Done